---
title: "Single Cell RNA-Seq : Coding Paper"
author: "Alexander G Foote"
date: "`r Sys.Date()`"
output:
    html_document:
      keep_md: TRUE
      toc: TRUE
---

# Single Cell RNA-Seq: Integration & Visualization 

```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE, error = FALSE)
```

## Load packkages

```{r workspace, include = FALSE}
library(Seurat)
library(kableExtra)
library(tidyr)
library(dplyr)
library(ggplot2)
library(ggsci)
library(scales)
library(HGNChelper)
library(ComplexHeatmap)
library(magrittr)
library(viridis)
library(dittoSeq)
library(data.table)
library(scCustomize)
library(qs)
library(paletteer)
library(igraph)
library(topGO)
library(enrichR)
library(org.Mm.eg.db)
library(biomaRt)
library(magrittr)
library(EnhancedVolcano)
set.seed(12345)
```

#### UMAP annotated clusters

```{r UMAP annotated clusters}

scRNA_epi_saline <- readRDS("/Users/alexanderfoote/dev/Projs/R/Upper_Airway_Coding_Project/rds/scRNA_epi_saline_rpca_refined_annotated_11122024.rds")
scRNA_epi_saline@meta.data

#Assign annotations for larynx vs tracheal regions
DefaultAssay(scRNA_epi_saline) <- "RNA"
Idents(scRNA_epi_saline) <- "seurat_clusters"
scRNA_epi_saline <- RenameIdents(scRNA_epi_saline, 
                   '0' =  'Tracheobronchial',
                  '1' =  'Tracheobronchial',
                  '2' =  'Tracheobronchial',
                  '3' =  'Tracheobronchial',
                  '4' =  'Tracheobronchial',
                  '11' =  'Tracheobronchial',
                  '18' = 'Tracheobronchial',
                   '15' = 'Tracheobronchial',
                  '16' =  'Submucosal-gland',
                  '9' = 'Submucosal-gland',
                  '13' = 'Submucosal-gland',
                  '7' = 'Submucosal-gland',
                  '10' = 'Submucosal-gland',
                  '5' = 'Submucosal-gland',
                  '12' =   'Pharyngolaryngeal',
                  '8' =   'Pharyngolaryngeal',
                  '14' =   'Pharyngolaryngeal',
                  '6' =   'Tracheobronchial',
                  '17' =  'Pharyngolaryngeal')

name <- "scRNA_epi_labels"
plot2 <- DimPlot(scRNA_epi_saline,reduction = "umap", pt.size=.1) + NoLegend()
LabelClusters(plot2, id="ident", size=5,repel=T, box.padding=.5)

#Assign cluster annotation for all cell types
scRNA_epi_saline$Region <- Idents(scRNA_epi_saline)

#Assign annotations to seurat_clusters
DefaultAssay(scRNA_epi_saline) <- "RNA"
Idents(scRNA_epi_saline) <- "seurat_clusters"
scRNA_epi_saline <- RenameIdents(scRNA_epi_saline, 
                  '0' =  'Basal-v.trachea',
                  '1' =  'Basal-d.trachea',
                  '12' =  'Basal-larynx',
                  '8' =  'Parabasal-larynx',
                  '14' =  'Suprabasal-larynx',
                  '6' =  'Secretory-trachea',
                  '17' =  'Secretory-larynx',
                  '16' =  'Basal-myoepithelial',
                  '2' =  'Club-proximal',
                  '3' =  'Club-mid',
                  '4' =  'Club-distal',
                  '11' = 'Club-proximal',
                  '9' =  'Serous-duct',
                  '5' = 'Goblet-2',
                  '10' = 'Serous-acini',
                  '7' = 'Goblet-2',
                  '13' = 'Goblet-1',
                  '18' =  'Tuft/NE/Ionocyte',
                  '15' = 'Ciliated')

name <- "scRNA_epi_labels"
plot1 <- DimPlot(scRNA_epi_saline,reduction = "umap", pt.size=.1) + NoLegend()
LabelClusters(plot1, id="ident", size=5,repel=T, box.padding=.5)

#Assign cluster annotation for all cell types
scRNA_epi_saline$CellType <- Idents(scRNA_epi_saline)
LabelClusters(plot1, id="ident", size=5,repel=T, box.padding=.5)
LabelClusters(plot2, id="ident", size=5,repel=T, box.padding=.5)
```

## Add custom cluster annotation

```{r}
NE <- WhichCells(object = scRNA_epi_saline, expression = Ascl1 > 1)
parentcluster <- WhichCells(object = scRNA_epi_saline, idents = 'Tuft/NE/Ionocyte')
NE <- NE[NE %in% parentcluster]
Idents(scRNA_epi_saline, cells = NE) <- 'Neuroendocrine'

Tuft <- WhichCells(object = scRNA_epi_saline, expression = Trpm5 > 1)
parentcluster <- WhichCells(object = scRNA_epi_saline, idents = 'Tuft/NE/Ionocyte')
Tuft <- Tuft[Tuft %in% parentcluster]
Idents(scRNA_epi_saline, cells = Tuft) <- 'Tuft'

#Reassign cluster annotation
scRNA_epi_saline$CellType <- Idents(scRNA_epi_saline)

scRNA_epi_saline$CellType <- factor(scRNA_epi_saline$CellType,levels=c('Basal-d.trachea','Basal-v.trachea','Basal-larynx','Parabasal-larynx','Basal-myoepithelial','Club-proximal','Club-mid','Club-distal',"Serous-duct",'Serous-acini','Goblet-1','Goblet-2','Secretory-unknown','Secretory-larynx','Suprabasal-larynx',"Neuroendocrine",'Tuft','Ciliated'))

scRNA_epi_saline$CellType <- factor(scRNA_epi_saline$CellType,levels=c('Basal-trachea','Basal-larynx','Parabasal-larynx','Basal-myoepithelial','Club-proximal','Club-mid','Club-distal',"Serous-duct",'Serous-acini','Goblet-1','Goblet-2','Secretory-trachea','Secretory-larynx','Suprabasal-larynx',"Neuroendocrine",'Tuft','Ciliated'))


#Remove 'Tuft/NE/Ionocyte' cluster
DimPlot(scRNA_epi_saline,reduction = "umap",label = TRUE,pt.size=1) 
#Reassign cluster annotation
scRNA_epi_saline$CellType <- Idents(scRNA_epi_saline)

name <- "scRNA_epi_labels_custom-annotate_celltype"
DimPlot_scCustom(scRNA_epi_saline,reduction = "umap",label = T,pt.size=.3, DiscretePalette_scCustomize(num_colors = 24, palette = "alphabet"), figure_plot = TRUE, label.size = 4, label.box = T) 

name <- "scRNA_epi_labels_custom-annotate_VlnPlotQC"
VlnPlot(scRNA_epi_saline, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3, pt.size = 0)
```

## Print all marker genes

```{r identify marker genes}
#Per Region
DefaultAssay(scRNA_epi_saline) <- "RNA"
Idents(scRNA_epi_saline) <- "Region"
markers.perregion <- FindAllMarkers(scRNA_epi_saline, only.pos = TRUE, min.pct = 0.25, thresh.use = 0.25)
tapply(markers.perregion$p_val_adj, markers.perregion$cluster, function(x){
  length(x < 0.05)
})
head(markers.perregion) %>%
  kable(table.attr = "style = \"color: black;\"",) %>%
  kable_styling("striped")
view.markers.perregion <- tapply(markers.perregion$gene, markers.perregion$cluster, function(x){head(x,1)})
# violin plots
lapply(view.markers.perregion, function(marker){
  VlnPlot_scCustom(scRNA_epi_saline,
          group.by = "Region",
          features = marker, pt.size=0)
})

# feature plots
lapply(view.markers.perregion, function(marker){
  FeaturePlot_scCustom(scRNA_epi_saline,
              features = marker)
})

name <- "epi_refined_rpca_annotated_782024_perregion_IAVseq.rds"
scRNA_epi_rpca.markers <- FindAllMarkers(scRNA_epi_saline, assay = "RNA", only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
x=scRNA_epi_rpca.markers %>% group_by(cluster) %>% top_n(n = 200, wt = avg_log2FC)
write.table(x, file=paste(name,"_Clustermarker.csv",sep=""), append = FALSE, quote = FALSE)

#Per CellType
DefaultAssay(scRNA_epi_saline) <- "RNA"
Idents(scRNA_epi_saline) <- "CellType"
markers.percell <- FindAllMarkers(scRNA_epi_saline, only.pos = TRUE, min.pct = 0.25, thresh.use = 0.25)
tapply(markers.percell$p_val_adj, markers.percell$cluster, function(x){
  length(x < 0.05)
})
head(markers.percell) %>%
  kable(table.attr = "style = \"color: black;\"",) %>%
  kable_styling("striped")
view.markers.percell <- tapply(markers.percell$gene, markers.percell$cluster, function(x){head(x,1)})
# violin plots
lapply(view.markers.percell, function(marker){
  VlnPlot_scCustom(scRNA_epi_saline,
          group.by = "CellType",
          features = marker, pt.size=0)
})
# feature plots
lapply(view.markers.percell, function(marker){
  FeaturePlot_scCustom(scRNA_epi_saline,
              features = marker)
})

name <- "epi.integrated_rpca_refined_salineonly_annotated_7122024_top500"
scRNA_epi_saline_rpca.markers <- FindAllMarkers(scRNA_epi_saline, assay = "RNA", only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
x=scRNA_epi_saline_rpca.markers %>% group_by(cluster) %>% top_n(n = 500, wt = avg_log2FC)
write.table(x, file=paste(name,"_Clustermarker.csv",sep=""), append = FALSE, quote = FALSE)
saveRDS(scRNA_epi_saline, file = "scRNA_epi_saline_rpca_refined_annotated_782024.rds")
```

## Visualize all marker genes on annotated seurat clusters per cell type

```{r plots of FindAllMarkers}
#Per Cell Type
DefaultAssay(scRNA_epi_saline) <- "RNA"
Idents(scRNA_epi_saline) <- "CellType"
scRNA_epi_saline$CellType <- factor(scRNA_epi_saline$CellType,levels=c('Basal-d.trachea','Basal-v.trachea','Basal-larynx','Parabasal-larynx','Basal-myoepithelial','Suprabasal-larynx','Secretory-trachea','Secretory-larynx','Club-proximal','Club-mid','Club-distal','Goblet-1','Goblet-2',"Serous-duct",'Serous-acini',"Neuroendocrine",'Tuft','Ciliated'))
celltype_markers <- c("Trp63","Cav1","Tgm2","Igfbp2","Tmprss11a","Acta2","Krt13","Kcnj16","Duoxa2","Slc6a15","Ces1f","Scgb1a1","Tff2","Lipf","Slc34a2","Bpifb5","Ascl1","Trpm5","Foxj1")
DotPlot_scCustom(scRNA_epi_saline, features=celltype_markers, colors_use= c("#E4E1E3FF","#66B0FFFF","#F6222EFF"), flip_axes = F, x_lab_rotate = TRUE)

# Find markers and limit to those expressed in greater than 60% of target population
all_markers <- FindAllMarkers(object = scRNA_epi_saline) %>%
    Add_Pct_Diff() %>%
    filter(pct_diff > 0.6)
top_markers <- Extract_Top_Markers(marker_dataframe = all_markers, num_genes = 5, named_vector = FALSE,
    make_unique = TRUE)
Clustered_DotPlot(scRNA_epi_saline, features = top_markers, k = 14)


DefaultAssay(scRNA_epi_saline) <- "integrated"
DoHeatmap(scRNA_epi_saline, angle = 90,size = 3, group.by = "CellType", features = celltype_markers) + scale_fill_gradientn(colors = c("#E4E1E3FF","#66B0FFFF","#F6222EFF"))

```

```{r Extensive dotplot}
#Per Cell Type
DefaultAssay(scRNA_epi_saline) <- "RNA"
Idents(scRNA_epi_saline) <- "CellType"
scRNA_epi_saline$CellType <- factor(scRNA_epi_saline$CellType,levels=c('Basal-d.trachea','Basal-v.trachea','Basal-larynx','Parabasal-larynx','Basal-myoepithelial','Suprabasal-larynx','Secretory-trachea','Secretory-larynx','Club-proximal','Club-mid','Club-distal','Goblet-1','Goblet-2','Serous-acini',"Serous-duct","Neuroendocrine",'Tuft','Ciliated'))
celltype_markers <- c("Trp63","Krt5","Krt17","Cav1","Tgm2","Dapk1","Lgr6","Igfbp2","Tmprss11a","Ntng1","Krt14","Ntrk3","Cxcl12","Acta2","Cxcl14","Krt13","Krt6a","Tmprss11b","Rbp2","Sprr1a","Muc1","Muc4","Muc20","Tnfaip2","Kcnj16","Tnfsf10","Il1a","Muc13","Bpifa1","Scgb1a1","Scgb3a2","Cyp2a5","Scgb3a1","Muc5b","Tff2","Lman1l","Wfdc12","Lipf","Dcpp3","Ltf","Dmbt1","Bpifb1","Bpifb5","Lyz1","Clu","Cryab","Slc34a2","Ascl1","Ngf","Cxcl13","Calca","Trpm5","Gnat3","Pou2f3","Lrmp","Dclk1","Spib","Foxj1","Ccdc39","Tmem212","Dynlrb2")
DotPlot_scCustom(scRNA_epi_saline, features=celltype_markers, colors_use= c("#E4E1E3FF","#66B0FFFF","#F6222EFF"), flip_axes = F, x_lab_rotate = TRUE)

DefaultAssay(scRNA_epi_saline) <- "integrated"
DoHeatmap(scRNA_epi_saline, angle = 90,size = 3, group.by = "CellType", features = celltype_markers) + scale_fill_gradientn(colors = c("#E4E1E3FF","#66B0FFFF","#F6222EFF"))

```

## Highlight a subset of cells

```{r highlight}
# Universal Step 1: List all identities and confirm they match the intended clusters
unique_clusters <- unique(Idents(scRNA_epi_saline))  # Check if this includes the desired identities

# Step 2: Create a named color vector FOR BASAL CELLS where each identity is explicitly matched to a color
# Assign grey to unhighlighted clusters by default
colors_use_basal <- rep("grey", length(unique_clusters))
names(colors_use_basal) <- unique_clusters

# Manually set colors for BASAL highlighted cell types
colors_use_basal["Basal-larynx"] <- "#993F00FF"
colors_use_basal["Basal-v.trachea"] <- "#0075DCFF"
colors_use_basal["Basal-d.trachea"] <- "#F0A0FFFF"
colors_use_basal["Basal-myoepithelial"] <- "#191919FF"

colors_use_basal["Basal-larynx"] <- "#993F00FF"
colors_use_basal["Basal-trachea"] <- "#F0A0FFFF"
colors_use_basal["Basal-myoepithelial"] <- "#191919FF"

# Step 3: Use DimPlot with this named color vector
# Re-order cell types
scRNA_epi_saline$CellType <- factor(scRNA_epi_saline$CellType,levels=c('Basal-v.trachea','Basal-d.trachea','Basal-larynx','Parabasal-larynx','Basal-myoepithelial','Suprabasal-larynx','Secretory-trachea','Secretory-larynx','Club-proximal','Club-mid','Club-distal','Goblet-1','Goblet-2',"Serous-duct",'Serous-acini',"Neuroendocrine",'Tuft','Ciliated'))

Idents(scRNA_epi_saline) <- "CellType"
DimPlot(
  scRNA_epi_saline, 
  group.by = "CellType",   # Ensure "CellType" is the correct metadata for grouping
  reduction = "umap",
  label = F, 
  pt.size = 1, 
  cols = colors_use_basal  # Named vector applies specific colors to specific identities
)

VlnPlot(
  scRNA_epi_saline, 
  features = c("Trp63"), 
  group.by = "CellType", 
  cols = colors_use_basal  # Set custom colors
)


# Step 2: Create a named color vector FOR SMG where each identity is explicitly matched to a color
# Assign grey to unhighlighted clusters by default
colors_use_SMG <- rep("grey", length(unique_clusters))
names(colors_use_SMG) <- unique_clusters

# Manually set colors for SMG highlighted cell types
colors_use_SMG["Serous-duct"] <- "#003380FF"
colors_use_SMG["Serous-acini"] <- "#FFA405FF"
colors_use_SMG["Goblet-1"] <- "#9DCC00FF"
colors_use_SMG["Goblet-2"] <- "#C20088FF"
colors_use_SMG["Basal-myoepithelial"] <- "#191919FF"

# Step 3: Use DimPlot with this named color vector
DimPlot(
  scRNA_epi_saline, 
  group.by = "CellType",   # Ensure "CellType" is the correct metadata for grouping
  reduction = "umap",
  label = F, 
  pt.size = 1, 
  cols = colors_use_SMG  # Named vector applies specific colors to specific identities
)


# Step 2: Create a named color vector FOR TERMINAL CELLS where each identity is explicitly matched to a color
# Assign grey to unhighlighted clusters by default
colors_use_ter <- rep("grey", length(unique_clusters))
names(colors_use_ter) <- unique_clusters

# Manually set colors for SMG highlighted cell types
colors_use_ter["Club-proximal"] <- "#808080FF" 
colors_use_ter["Club-mid"] <- "#94FFB5FF"
colors_use_ter["Club-distal"] <-  "#8F7C00FF"
colors_use_ter["Ciliated"] <- "#FF0010FF" 
colors_use_ter["Neuroendocrine"] <- "#FFA8BBFF"
colors_use_ter["Tuft"] <- "#005C31FF"
colors_use_ter["Suprabasal-larynx"] <- "#426600FF"
colors_use_ter["Secretory-trachea"] <- "#2BCE48FF"
colors_use_ter["Secretory-larynx"] <- "#FFCC99FF" 

# Step 3: Use DimPlot with this named color vector
DimPlot(
  scRNA_epi_saline, 
  group.by = "CellType",   # Ensure "CellType" is the correct metadata for grouping
  reduction = "umap",
  label = F, 
  pt.size = 1, 
  cols = colors_use_ter  # Named vector applies specific colors to specific identities
)


# Step 2: Create a named color vector FOR LARYNX CELLS where each identity is explicitly matched to a color
# Assign grey to unhighlighted clusters by default
colors_use_larynx <- rep("grey", length(unique_clusters))
names(colors_use_larynx) <- unique_clusters

# Manually set colors for SMG highlighted cell types
colors_use_larynx["Basal-larynx"] <- "#993F00FF"
colors_use_larynx["Parabasal-larynx"] <- "#FFCC99FF" 
colors_use_larynx["Suprabasal-larynx"] <- "#4C005CFF" 
colors_use_larynx["Secretory-larynx"] <- "#005C31FF"

# Step 3: Use DimPlot with this named color vector
DimPlot(
  scRNA_epi_saline, 
  group.by = "CellType",   # Ensure "CellType" is the correct metadata for grouping
  reduction = "umap",
  label = F, 
  pt.size = 1, 
  cols = colors_use_larynx  # Named vector applies specific colors to specific identities
)


# Display the "alphabet" palette using show_palette() from ggsci
colors <- DiscretePalette_scCustomize(n = 24, palette = "alphabet")
# Use show_col to visualize the colors
show_col(colors)
#FOAOFFFF#0075DCFF#993F00FF#4C005CFF#191919FF
#005C31FF#2BCE48F#FFCC99F#808080FF#94FFB5FF
#8F7C00F#9DCC00FF#C20088FF#003380FF#FFA405FF
#FFA8BBF#426600FF#FF0010FF#5EF1F2FF#00998FFF
#E0FF66FF#740AFFFF#990000FF#FFFF80FF
```

## Investigate the relationship between cluster identity and sample identity. Building a phylogenetic tree relating the 'average' cell from each group. This tree is estimated based on a distance matrix constructed in either gene expression space or PCA space.

```{r membership}
#Per CELL TYPE
head(scRNA_epi_saline@meta.data)
Idents(scRNA_epi_saline) <- scRNA_epi_saline$CellType
## extract meta data
membership <- scRNA_epi_saline@meta.data %>% as.data.table # the resulting membership object has one "row" per cell
## count the number of cells per unique combinations of "Condition" and "CellType"
membership[, .N, by = c("Condition", "CellType")]
## with additional casting after the counting
membership[, .N, by = c("Condition", "CellType")] %>% dcast(., Condition ~ CellType, value.var = "N")
  
dittoBarPlot(scRNA_epi_saline, "Condition", group.by = "CellType",
             main = "Cluster Composition",
             y.breaks = c(0,.1,.2,.3,.4,.5,.6,.7,.8,.9,1),
             xlab = NULL, # NULL = remove
             ylab = "% of total",
              color.panel = c("skyblue","brown"))
  
scRNA_epi_saline <- BuildClusterTree(scRNA_epi_saline, dims = 1:50)
PlotClusterTree(scRNA_epi_saline)


Total_epi_cell_count <- ggplot(scRNA_epi_saline@meta.data, aes(CellType)) +
  geom_bar(stat="count", fill='red', colour='black', size = 0.3, width = 0.8, key_glyph = draw_key_label, alpha=.6) +
labs(title = NULL, subtitle=NULL, x = NULL, y = 'Total Cell Count', caption = NULL, fill = NULL) + 
  theme(legend.text = element_text(size = 15, face=NULL)) +
  theme(plot.title=element_text(size=13, face="bold"), 
        axis.text.x=element_text(size=12, face="bold", angle = 40, hjust = 1), 
        axis.text.y=element_text(size=7),
        axis.title.x=element_text(size=8, face=NULL),
        axis.title.y=element_text(size =14)) +
        scale_y_continuous(expand = c(0, 0), breaks = seq(0, 1800, 100)) +
    coord_cartesian(ylim=c(0, 1800)) +
  theme(panel.grid = element_blank(), panel.background = element_blank(), axis.line = element_line(colour = 'black'), legend.position="right") 
VlnPlot_scCustom(scRNA_epi_saline, features = "nCount_RNA", split.by = "Condition", group.by = "CellType", pt.size = 0)
```

#Explore DEG with advanced visualization

```{r identify unique cell clusters}
#Define assay
DefaultAssay(scRNA_epi_saline) <- "RNA"
#SMG subpopulations
name <- "SMG_populations"
FeaturePlot_scCustom(scRNA_epi_saline, 
            features = c("Dmbt1","Acta2","Nkx3-1","Muc5b","Tff2","Lipf","Slc34a2","Bpifb5"), 
            order=TRUE, pt.size=1, colors_use= c("#E4E1E3FF","#F6222EFF"), na_cutoff =.5, num_columns = 4)

FeaturePlot_scCustom(scRNA_epi_saline, 
            features = c("Dmbt1"), 
            order=TRUE, pt.size=2, colors_use= c("#E4E1E3FF","#F6222EFF"), na_cutoff =2)

FeaturePlot_scCustom(scRNA_epi_saline, 
            features = c("Acta2"), 
            order=TRUE, pt.size=2, colors_use= c("#E4E1E3FF","#F6222EFF"), na_cutoff =1)

FeaturePlot_scCustom(scRNA_epi_saline, 
            features = c("Nkx3-1"), 
            order=TRUE, pt.size=2, colors_use= c("#E4E1E3FF","#F6222EFF"), na_cutoff =.5)

FeaturePlot_scCustom(scRNA_epi_saline, 
            features = c("Muc5b"), 
            order=TRUE, pt.size=2, colors_use= c("#E4E1E3FF","#F6222EFF"), na_cutoff =3)

FeaturePlot_scCustom(scRNA_epi_saline, 
            features = c("Tff2"), 
            order=TRUE, pt.size=2, colors_use= c("#E4E1E3FF","#F6222EFF"), na_cutoff =3)

FeaturePlot_scCustom(scRNA_epi_saline, 
            features = c("Lipf"), 
            order=TRUE, pt.size=2, colors_use= c("#E4E1E3FF","#F6222EFF"), na_cutoff =4)

FeaturePlot_scCustom(scRNA_epi_saline, 
            features = c("Slc34a2"), 
            order=TRUE, pt.size=2, colors_use= c("#E4E1E3FF","#F6222EFF"), na_cutoff =.5)

FeaturePlot_scCustom(scRNA_epi_saline, 
            features = c("Bpifb5"), 
            order=TRUE, pt.size=2, colors_use= c("#E4E1E3FF","#F6222EFF"), na_cutoff =.5)

FeaturePlot_scCustom(scRNA_epi_saline, 
            features = c("Nkx3-1","Sox9","Bpifb2","Azgp1"), 
            order=TRUE, pt.size=2, colors_use= c("#E4E1E3FF","#F6222EFF"), na_cutoff =.5)


#Basal-to-luminal markers
FeaturePlot_scCustom(scRNA_epi_saline, 
            features = c("Igfbp2"), 
            order=TRUE,pt.size=2, colors_use= c("#E4E1E3FF","#F6222EFF"), na_cutoff =1.5)

FeaturePlot_scCustom(scRNA_epi_saline, 
            features = c("Tmprss11a"), 
            order=TRUE,pt.size=2, colors_use= c("#E4E1E3FF","#F6222EFF"), na_cutoff =1.5)

FeaturePlot_scCustom(scRNA_epi_saline, 
            features = c("Krt13"), 
            order=TRUE,pt.size=2, colors_use= c("#E4E1E3FF","#F6222EFF"), na_cutoff =4)


#Regional tissue markers
FeaturePlot_scCustom(scRNA_epi_saline, 
            features = c("Tmprss11a"), 
            order=TRUE,pt.size=2, colors_use= c("#E4E1E3FF","#F6222EFF"), na_cutoff =1.5)

FeaturePlot_scCustom(scRNA_epi_saline, 
            features = c("Nkx2-1"), 
            order=TRUE,pt.size=2, colors_use= c("#E4E1E3FF","#F6222EFF"), na_cutoff =.5)

FeaturePlot_scCustom(scRNA_epi_saline, 
            features = c("Dmbt1"), 
            order=TRUE,pt.size=2, colors_use= c("#E4E1E3FF","#F6222EFF"), na_cutoff =2)

FeaturePlot_scCustom(scRNA_epi_saline, 
            features = c("Sox9"), 
            order=TRUE,pt.size=2, colors_use= c("#E4E1E3FF","#F6222EFF"), na_cutoff =.5)


#Club markers
FeaturePlot_scCustom(scRNA_epi_saline, 
            features = c("Scgb1a1"), 
            order=TRUE,pt.size=1, colors_use= c("#E4E1E3FF","#F6222EFF"), na_cutoff =4)

FeaturePlot_scCustom(scRNA_epi_saline, 
            features = c("Scgb3a2"), 
            order=TRUE,pt.size=1, colors_use= c("#E4E1E3FF","#F6222EFF"), na_cutoff =4)


#Keratin markers
FeaturePlot_scCustom(scRNA_epi_saline, 
            features = c('Krt1', 'Krt2', 'Krt3', 'Krt4', 'Krt5', 'Krt6a', 'Krt6b', 'Krt6c', 'Krt7', 'Krt8', 'Krt9', 'Krt10', 'Krt12', 'Krt13', 'Krt14', 'Krt15', 'Krt16', 'Krt17', 'Krt18', 'Krt19', 'Krt20', 'Krt21', 'Krt23', 'Krt24', 'Krt25', 'Krt26', 'Krt27', 'Krt28', 'Krt31', 'Krt32', 'Krt33a', 'Krt33b', 'Krt34', 'Krt35', 'Krt36', 'Krt37', 'Krt38', 'Krt39', 'Krt40', 'Krt71', 'Krt72', 'Krt73', 'Krt74', 'Krt75', 'Krt76', 'Krt77', 'Krt78', 'Krt79', 'Krt80', 'Krt81', 'Krt82', 'Krt83', 'Krt84', 'Krt85', 'Krt86', 'Krt87', 'Krt88', 'Krt89', 'Krt90', 'Krt93', 'Krt94', 'Krt95', 'Krt96'), 
            colors_use= c("#E4E1E3FF","#F6222EFF"), na_cutoff =.5, num_columns = 3)

DotPlot_scCustom(scRNA_epi_saline, 
            features = c('Krt1', 'Krt2', 'Krt3', 'Krt4', 'Krt5', 'Krt6a', 'Krt6b', 'Krt6c', 'Krt7', 'Krt8', 'Krt9', 'Krt10', 'Krt12', 'Krt13', 'Krt14', 'Krt15', 'Krt16', 'Krt17', 'Krt18', 'Krt19', 'Krt20', 'Krt21', 'Krt23', 'Krt24', 'Krt25', 'Krt26', 'Krt27', 'Krt28', 'Krt31', 'Krt32', 'Krt33a', 'Krt33b', 'Krt34', 'Krt35', 'Krt36', 'Krt37', 'Krt38', 'Krt39', 'Krt40', 'Krt71', 'Krt72', 'Krt73', 'Krt74', 'Krt75', 'Krt76', 'Krt77', 'Krt78', 'Krt79', 'Krt80', 'Krt81', 'Krt82', 'Krt83', 'Krt84', 'Krt85', 'Krt86', 'Krt87', 'Krt88', 'Krt89', 'Krt90', 'Krt93', 'Krt94', 'Krt95', 'Krt96'), x_lab_rotate = TRUE,
            colors_use= c("#E4E1E3FF","#F6222EFF")) #63 known KRTs

```

#### Session information

```{r sessioinfo}
sessionInfo()
```
